### PR TITLE
DSS read-only mode

### PR DESCRIPTION
**Pull request by xbrianh**
For DSS write paths in the API server, check if `DSS_READ_ONLY_MODE` is defined in the environment. If it is, return 405.

### CODE REVIEW

File: dss/error.py [hunk 1]
```diff
@@ -42,27 +43,41 @@ def dss_exception_handler(e: DSSException) -> FlaskResponse:
             'stacktrace': traceback.format_exc(),
         })
 
-
 def dss_handler(func):
```
[the remark this answers is not in the document]
Comment (xbrianh) [reply dss/error.py]: It's still used in `dss.api.search`, in a `POST` method. This is ambiguous to me, but I guess it's a read path? If it is we can dump `dss_handler`, what do you think?

File: tests/test_api.py [hunk 2]
```diff
@@ -66,6 +68,33 @@ def test_get_version(self):
...
+                            es_query={},
+                            callback_url="https://take.me.to.funkytown",
+                            source_url="s3://urlurlurlbaby",
+                            description="supercalifragilisticexpialidocious",
+                            details={},
+                            contents=[]
+                        )
+                    )
+
+        del os.environ['DSS_READ_ONLY_MODE']
```
[the remark this answers is not in the document]
Comment (xbrianh) [reply tests/test_api.py]: Will fix

File: dss/error.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ttung) [reply dss/error.py]: I think that should be a read path.

**Comment by Bento007** [comment]
Why did code coverage drop so much?

**Comment by ttung** [comment]
If you want to be crazy ambitious, you can proxy cloud-blobstore and the elasticsearch libraries to trap writes. :)

**Comment by kislyuk** [comment]
I agree with @Bento007, it would be nice to simplify this if possible. The `POST /search` can be treated as an exception.

File: dss/error.py [hunk 3]
```diff
@@ -42,27 +44,34 @@ def dss_exception_handler(e: DSSException) -> FlaskResponse:
...
+                stacktrace = traceback.format_exc()
+            except Exception as ex:
+                status = requests.codes.server_error
+                code = "unhandled_exception"
+                title = str(ex)
+                stacktrace = traceback.format_exc()
+        else:
+            status = requests.codes.not_allowed
+            code = "read_only"
+            title = "The DSS is currently read-only"
```
Comment (kislyuk) [inline dss/error.py]: Consider changing this to a 5xx series error code, and adding a Retry-After header.

**Pull request merged**